composeApp/src/androidMain/kotlin/sp/bvantur/tasky/core/data/TaskySyncScheduler.android.kt

```diff
@@ -3,41 +3,84 @@ package sp.bvantur.tasky.core.data
+import androidx.work.OneTimeWorkRequest
 import androidx.work.OneTimeWorkRequestBuilder
 import androidx.work.WorkManager
 import kotlinx.coroutines.CoroutineScope
 import kotlinx.coroutines.launch
+import sp.bvantur.tasky.core.data.local.EventEntity
+import sp.bvantur.tasky.core.data.local.SyncStep
 import java.util.concurrent.TimeUnit
 
 actual class TaskySyncScheduler(
     private val workManager: WorkManager,
     private val applicationCoroutineScope: CoroutineScope
 ) {
-    actual suspend fun scheduleAgendaSync(eventIds: List<String>) {
-        eventIds.forEach { eventId ->
-            val agendaWorker = OneTimeWorkRequestBuilder<SyncEventsWorker>()
-                .setConstraints(
-                    Constraints.Builder()
-                        .setRequiredNetworkType(NetworkType.CONNECTED)
-                        .build()
-                )
-                .setBackoffCriteria(
-                    backoffPolicy = BackoffPolicy.EXPONENTIAL,
-                    backoffDelay = 5000,
-                    timeUnit = TimeUnit.MILLISECONDS
-                )
-                .addTag("add_event")
-                .setInputData(
-                    Data.Builder()
-                        .putString(SyncEventsWorker.SYNC_EVENT_ID, eventId)
-                        .build()
-                )
-                .build()
+    actual suspend fun scheduleAgendaSync(events: List<EventEntity>) {
+        events.forEach { event ->
+            val agendaWorker: OneTimeWorkRequest = when (event.syncStep) {
+                SyncStep.CREATE -> {
+                    onCreateEventWorker(event)
+                }
+                SyncStep.EDIT -> {
+                    return@forEach // TODO
+                }
+                SyncStep.DELETE -> {
+                    onDeleteEventWorker(event)
+                }
+
+                else -> return@forEach
+            }
 
             applicationCoroutineScope.launch {
-                workManager.enqueue(agendaWorker)
+                workManager.beginUniqueWork(
+                    event.id,
+                    ExistingWorkPolicy.REPLACE,
+                    agendaWorker
+                ).enqueue()
             }
         }
     }
+
+    private fun onCreateEventWorker(event: EventEntity): OneTimeWorkRequest =
+        OneTimeWorkRequestBuilder<SyncCreateEventWorker>()
+            .setConstraints(
+                Constraints.Builder()
+                    .setRequiredNetworkType(NetworkType.CONNECTED)
+                    .build()
+            )
+            .setBackoffCriteria(
+                backoffPolicy = BackoffPolicy.EXPONENTIAL,
+                backoffDelay = 5000,
+                timeUnit = TimeUnit.MILLISECONDS
+            )
+            .addTag("add_event")
+            .setInputData(
+                Data.Builder()
+                    .putString(SyncCreateEventWorker.SYNC_EVENT_ID, event.id)
+                    .build()
+            )
+            .build()
```

Comment:
After creating a bunch of these workers, I made these extension functions, which I think you'll find helpful 😊

```
fun OneTimeWorkRequest.Builder.setRequireNetworkConnectivity(): OneTimeWorkRequest.Builder {
    return this.setConstraints(
        Constraints.Builder()
            .setRequiredNetworkType(NetworkType.CONNECTED)
            .build()
    )
}

fun OneTimeWorkRequest.Builder.setExponentialBackOff(duration: Long): OneTimeWorkRequest.Builder {
    return this.setBackoffCriteria(
        backoffPolicy = BackoffPolicy.EXPONENTIAL,
        backoffDelay = duration,
        timeUnit = TimeUnit.MILLISECONDS
    )
}

fun OneTimeWorkRequest.Builder.setInputParameters(block: Data.Builder.() -> Unit): OneTimeWorkRequest.Builder {
    return this.setInputData(
        Data.Builder()
            .apply(block)
            .build()
    )
}
```

Usage looks like this, which I found easy to read and understand:

```
val workRequest = OneTimeWorkRequestBuilder<DeleteReminderWorker>()
  .addTag(DeleteReminderWorker.TAG)
  .setRequireNetworkConnectivity()
  .setExponentialBackOff(2000L)
  .setInputParameters { putString(DeleteReminderWorker.REMINDER_ID, reminderId) }
  .build()
```